### PR TITLE
Fix ICE with `ReadPointerAsBytes` validation error

### LINKED ISSUE #79690: ICE: compiler/rustc_mir/src/interpret/validity.rs:918:17: Unexpected error during validation: unable to turn pointer into raw bytes
### Code

```Rust
union Transmute<T: Copy, U: Copy> {
    t: T,
    u: U,
}
trait Bar {
    fn bar(&self) -> u32;
}
struct Foo {
    foo: u32,
    bar: bool,
}
impl Bar for Foo {
    fn bar(&self) -> u32 {
        self.foo
    }
}
#[derive(Copy, Clone)]
struct Fat<'a>(&'a Foo, &'static VTable);
struct VTable {
    size: Foo,
}
const FOO: &Bar = &Foo {
    foo: 128,
    bar: false,
};
const G: Fat = unsafe { Transmute { t: FOO }.u };
```

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (5be3f9f10 2020-12-03)
binary: rustc
commit-hash: 5be3f9f10e9fd59ea03816840a6051413fbdefae
commit-date: 2020-12-03
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

### Error output

```
warning: trait objects without an explicit `dyn` are deprecated
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:22:13
   |
22 | const FOO: &Bar = &Foo {
   |             ^^^ help: use `dyn`: `dyn Bar`
   |
   = note: `#[warn(bare_trait_objects)]` on by default

warning: union is never used: `Transmute`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:7
  |
1 | union Transmute<T: Copy, U: Copy> {
  |       ^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: field is never read: `bar`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:10:5
   |
10 |     bar: bool,
   |     ^^^^^^^^^

warning: field is never read: `size`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:20:5
   |
20 |     size: Foo,
   |     ^^^^^^^^^

warning: constant is never used: `FOO`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:22:1
   |
22 | / const FOO: &Bar = &Foo {
23 | |     foo: 128,
24 | |     bar: false,
25 | | };
   | |__^

warning: constant is never used: `G`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:26:1
   |
26 | const G: Fat = unsafe { Transmute { t: FOO }.u };
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: compiler/rustc_mir/src/interpret/validity.rs:918:17: Unexpected error during validation: unable to turn pointer into raw bytes

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:958:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0-nightly (5be3f9f10 2020-12-03) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type staticlib

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `G`
#1 [eval_to_const_value_raw] simplifying constant for the type system `G`
end of query stack
error: aborting due to previous error; 6 warnings emitted
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
warning: trait objects without an explicit `dyn` are deprecated
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:22:13
   |
22 | const FOO: &Bar = &Foo {
   |             ^^^ help: use `dyn`: `dyn Bar`
   |
   = note: `#[warn(bare_trait_objects)]` on by default

warning: union is never used: `Transmute`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:7
  |
1 | union Transmute<T: Copy, U: Copy> {
  |       ^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: field is never read: `bar`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:10:5
   |
10 |     bar: bool,
   |     ^^^^^^^^^

warning: field is never read: `size`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:20:5
   |
20 |     size: Foo,
   |     ^^^^^^^^^

warning: constant is never used: `FOO`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:22:1
   |
22 | / const FOO: &Bar = &Foo {
23 | |     foo: 128,
24 | |     bar: false,
25 | | };
   | |__^

warning: constant is never used: `G`
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:26:1
   |
26 | const G: Fat = unsafe { Transmute { t: FOO }.u };
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: compiler/rustc_mir/src/interpret/validity.rs:918:17: Unexpected error during validation: unable to turn pointer into raw bytes

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:958:9
stack backtrace:
   0: std::panicking::begin_panic
   1: rustc_errors::HandlerInner::bug
   2: rustc_errors::Handler::bug
   3: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
   4: rustc_middle::ty::context::tls::with_opt::{{closure}}
   5: rustc_middle::ty::context::tls::with_opt
   6: rustc_middle::util::bug::opt_span_bug_fmt
   7: rustc_middle::util::bug::bug_fmt
   8: rustc_mir::interpret::validity::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::validate_operand_internal
   9: rustc_mir::const_eval::eval_queries::eval_to_allocation_raw_provider
  10: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::eval_to_allocation_raw>::compute
  11: rustc_middle::ty::query::plumbing::<impl rustc_query_system::query::QueryContext for rustc_middle::ty::context::TyCtxt>::start_query::{{closure}}::{{closure}}::{{closure}}
  12: rustc_query_system::query::plumbing::get_query_impl
  13: rustc_mir::const_eval::eval_queries::eval_to_const_value_raw_provider
  14: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::eval_to_const_value_raw>::compute
  15: rustc_middle::ty::query::plumbing::<impl rustc_query_system::query::QueryContext for rustc_middle::ty::context::TyCtxt>::start_query::{{closure}}::{{closure}}::{{closure}}
  16: rustc_query_system::query::plumbing::get_query_impl
  17: rustc_mir::const_eval::eval_queries::eval_to_const_value_raw_provider
  18: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::eval_to_const_value_raw>::compute
  19: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task
  20: rustc_data_structures::stack::ensure_sufficient_stack
  21: rustc_query_system::query::plumbing::get_query_impl
  22: rustc_middle::mir::interpret::queries::<impl rustc_middle::ty::context::TyCtxt>::const_eval_poly
  23: <rustc_lint::BuiltinCombinedLateLintPass as rustc_lint::passes::LateLintPass>::check_item
  24: rustc_hir::intravisit::Visitor::visit_nested_item
  25: rustc_hir::intravisit::walk_crate
  26: rustc_session::utils::<impl rustc_session::session::Session>::time
  27: rustc_interface::passes::analysis::{{closure}}::{{closure}}
  28: rustc_session::utils::<impl rustc_session::session::Session>::time
  29: rustc_interface::passes::analysis
  30: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  31: rustc_query_system::dep_graph::graph::DepGraph<K>::with_eval_always_task
  32: rustc_data_structures::stack::ensure_sufficient_stack
  33: rustc_query_system::query::plumbing::get_query_impl
  34: rustc_interface::passes::QueryContext::enter
  35: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  36: rustc_span::with_source_map
  37: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0-nightly (3ff10e74a 2020-12-04) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type staticlib

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `G`
#1 [eval_to_const_value_raw] simplifying constant for the type system `G`
#2 [eval_to_const_value_raw] simplifying constant for the type system `G`
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error; 6 warnings emitted
```

</p>
</details>

### PR DESCRIPTION
**Pull request by camelid**
Fixes #79690.

r? @oli-obk

### CODE REVIEW

File: src/test/ui/issues/issue-79690.stderr [hunk 1]
```diff
@@ -0,0 +1,11 @@
+error[E0080]: it is undefined behavior to use this value
+  --> $DIR/issue-79690.rs:26:1
+   |
+LL | const G: Fat = unsafe { Transmute { t: FOO }.u };
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered read of part of a pointer at .1.<deref>.size.foo
```
Comment (camelid) [inline src/test/ui/issues/issue-79690.stderr:5]: This error message doesn't seem super clear to me. In fact, I'm not sure if *I* understand it! Will it be understandable to people getting this error, or can it be improved somehow?

**Comment by camelid** [comment]
Huh, I guess this test made us reach tidy's limit!

```
tidy error: following path contains more than 2830 entries, you should move the test to some relevant subdirectory (current: 2830): /checkout/src/test/ui/issues
```

What should I do?

**Comment by JohnTitor** [comment]
> 
> 
> Huh, I guess this test made us reach tidy's limit!
> 
> ```
> tidy error: following path contains more than 2830 entries, you should move the test to some relevant subdirectory (current: 2830): /checkout/src/test/ui/issues
> ```
> 
> What should I do?

I guess you're the first case? 😄 You could move it to `src/test/ui/consts`.

**Comment by camelid** [comment]
Um... that's really weird: I moved the test and the error message changed:

```diff
 ---- [ui] ui/consts/issue-79690.rs stdout ----
diff of stderr:

2	  --> $DIR/issue-79690.rs:26:1
3	   |
4	LL | const G: Fat = unsafe { Transmute { t: FOO }.u };
-	   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered read of part of a pointer at .1.<deref>.size.foo
+	   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered pointer to alloc4 at .1.<deref>.size.foo, but expected initialized plain (non-pointer) bytes
6	   |
7	   = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
8	
```

**Comment by oli-obk** [comment]
That's super odd. I don't think it can be the move, maybe something changed on master under you? Try rebasing and reblessing.

**Comment by camelid** [comment]
Rebased. Rerunning test locally.

**Comment by camelid** [comment]
Huh, the test passed locally but not on CI...

**Comment by camelid** [comment]
> That's super odd. I don't think it can be the move, maybe something changed on master under you? Try rebasing and reblessing.

@oli-obk I rebased and ran the test locally, and it passed. But, it's still failing on CI.

- `git status` doesn't show any difference between my local repo and the one on GitHub
- I ran my stage1 `rustc` manually, and the error looks correct
- The merge-base is 5e91c4ecc09312d8b63d250a432b0f3ef83f1df7, which is from just earlier today

So, I'm not sure what's going on. Maybe we should get someone from T-infra to take a look at it?

**Comment by camelid** [comment]
The one thought I have is that maybe this is some kind of platform-specific code in the compiler that's being executed (I'm on macOS and CI is Linux, so it's different platforms), but I don't see why this compiler code should be platform-specific.

**Comment by oli-obk** [comment]
Aaah... I found the issue. CI is testing on 32 bit. So just ignore the test on 32 bit, it's not worth supporting that case in this specific test.

**Comment by camelid** [comment]
> Aaah... I found the issue. CI is testing on 32 bit. So just ignore the test on 32 bit, it's not worth supporting that case in this specific test.

Done!

**Comment by camelid** [comment]
(I think `@bors squash` doesn't work in rollups though, so I can squash manually if you like.)

**Comment by oli-obk** [comment]
@bors r+ rollup

:shrug: it's fine by me

**Pull request merged**